### PR TITLE
Stop changing created_at

### PR DESCRIPTION
**Pull request by musialik**
## Ticket
https://netguru.atlassian.net/browse/RD-97

## Description
- All logic related to project check deadlines now lives in `DueDatePolicy`
- The policy object has tests for edge cases
- A feature spec for due dates was added to make sure the policy object is used properly
- Enabling/disabling a project check and overriding its deadline no longer change the `created_at` column.

### CODE REVIEW

File: app/policies/due_date_policy.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+  private
+
+  def date_to_count_from
+    if checked_before?
+      [project_check.last_check_date,
+       project_check.last_check_date_without_disabled_period].compact.max
+    else
+      project_check.created_at_without_disabled_period || project_check.created_at
```
Comment (maciekkolodziej) [inline app/policies/due_date_policy.rb]: I thought that we wanted to leave `created_at` alone ;-)

File: app/policies/due_date_policy.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
+class DueDatePolicy
```
Comment (jedi1156) [inline app/policies/due_date_policy.rb]: Which of the public methods of this class are used outside of this class?

File: spec/policies/due_date_policy_spec.rb [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        context "and was disabled for 7 days" do
+          before { disable_check_for(7.days) }
+
+          it { is_expected.to eq(2.days.from_now.to_date) }
+        end
+      end
+    end
+  end
+
+  describe "#due_in" do
```
Comment (jedi1156) [inline spec/policies/due_date_policy_spec.rb]: I suggest:
- using `subject { policy.due_in }`
- using 3 contexts here - one for each example
- in each context a `let` for date returned by `due_on`
- extracting the stub to `before` for all contexts
```
describe "#due_in" do
  subject { policy.due_in }
  before do
    allow(subject).to receive(:due_on).and_return(due_on_date)
  end

  context "when due on date is in the future" do
    let(:due_on_date) { 20.days.from_now.to_date }

    it "returns the difference between today and due date in days"
      is_expected.to eq 20
    end
  end

  context "when due on date is in the past" do
    let(:due_on_date) { 20.days.ago.to_date }

    it "returns the negative difference between today and due date in days"
      is_expected.to eq -20
    end
  end

  context "when due on date is today" do
    let(:due_on_date) { Time.zone.today.to_date }

    it { is_expected.to eq 0 }
  end
end
```

File: spec/policies/due_date_policy_spec.rb [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+      expect(subject.due_in).to eq(-20)
+    end
+
+    it "equals zero if due today" do
+      allow(subject).to receive(:due_on).and_return(Time.zone.today.to_date)
+      expect(subject.due_in).to eq(0)
+    end
+  end
+
+  describe "#overdue?" do
```
Comment (jedi1156) [inline spec/policies/due_date_policy_spec.rb]: I suggest applying the same scheme to this `describe` as for "#due_in". Here we have the same contexts - they can be shared in this spec file.

File: spec/policies/due_date_policy_spec.rb [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+  describe "#remind_on" do
+    subject { policy.remind_on }
+    let(:remind_after_days) { [5, 30, 45] }
+    let(:init_remind_after_days) { [2, 6, 8] }
+
+    context "when never checked" do
+      let(:check) { create_check(created_at: 4.days.ago) }
+
+      it "calculates notification dates from created_at" do
+        expect(subject).to eq([2.days.ago,
```
Comment (jedi1156) [inline spec/policies/due_date_policy_spec.rb]: `expect(subject)` === `is_expected`

File: spec/policies/due_date_policy_spec.rb [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+  it { is_expected.to respond_to(:project_check) }
+
+  describe "#due_on" do
+    subject { policy.due_on }
+
+    context "when configured to be valid for 60 days" do
+      let(:valid_for_n_days) { 60 }
+
+      context "and was checked 70 days ago" do
+        let(:check) { create_check(last_check_date: 70.days.ago) }
```
Comment (jedi1156) [inline spec/policies/due_date_policy_spec.rb]: This check is not created for the example. Use `let!` instead.

File: spec/policies/due_date_policy_spec.rb [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+          it { is_expected.to eq(10.days.from_now.to_date) }
+        end
+      end
+    end
+
+    context "when configured to be first checked within 10 days" do
+      let(:init_valid_for_n_days) { 10 }
+
+      context "and was created 15 days ago and never checked" do
+        let(:check) { create_check(created_at: 15.days.ago) }
```
Comment (jedi1156) [inline spec/policies/due_date_policy_spec.rb]: The same for the other `check`s - I don't think they are created at all.

File: spec/policies/due_date_policy_spec.rb [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+  let(:valid_for_n_days) { 90 }
+  let(:remind_after_days) { [30, 60, 85] }
+  let(:init_valid_for_n_days) { 10 }
+  let(:init_remind_after_days) { [5, 8] }
+
+  def create_check(attrs = {})
+    create(:project_check, attrs.merge(reminder: reminder))
+  end
+
+  def disable_check_for(duration)
```
Comment (jedi1156) [inline spec/policies/due_date_policy_spec.rb]: Enclosed in `before` it can be a `shared_context` with a parameter. I am not a fan of methods in specs, but it is no sin ;) just a suggestion

File: spec/policies/due_date_policy_spec.rb [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+      it "includes the disabled period in calculations" do
+        expect(subject).to eq([25.days.ago,
+                               Time.zone.today,
+                               15.days.from_now].map(&:to_date))
+      end
+    end
+  end
+
+  describe "#remind_on?" do
+    it "checks if a reminder should be sent on given date" do
```
Comment (jedi1156) [inline spec/policies/due_date_policy_spec.rb]: Use `:aggregate_failures` for tests where you have multiple expectations. Still, here you can have a context and two separate tests.

File: spec/policies/due_date_policy_spec.rb [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+      allow(subject).to receive(:remind_on)
```
Comment (jedi1156) [inline spec/policies/due_date_policy_spec.rb]: I suggest extracting this stub the same way as for `#due_in`

File: app/jobs/project_checked_on_time_job.rb [hunk 11]
```diff
@@ -8,18 +8,23 @@ def initialize(project_check_id)
 
   # rubocop:disable Metrics/AbcSize
```
Comment (jedi1156) [inline app/jobs/project_checked_on_time_job.rb]: Is there a reason for disabling this cop here?

File: spec/policies/due_date_policy_spec.rb [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    end
+  end
+
+  describe "#remind_today?" do
+    context "if reminder should be sent today" do
+      it "returns true" do
+        allow(subject).to receive(:remind_on)
+          .and_return([Time.zone.today.to_date])
+
+        expect(subject.remind_today?).to eq true
```
Comment (jedi1156) [inline spec/policies/due_date_policy_spec.rb]: Having the subject as `subject { policy.remind_today? }` and the stub extracted to before you can write:
```
it { is_expected.to be true }
```

File: spec/features/due_dates_spec.rb [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+  end
+
+  def reload!
+    reminder_page.load reminder_id: reminder.id
+  end
+
+  context "A project_check that was never checked" do
+    scenario "and isn't overdue" do
+      Timecop.travel(10.days.from_now)
+      reminder_page.load reminder_id: reminder.id
```
Comment (jedi1156) [inline spec/features/due_dates_spec.rb]: There are no assertions for this page load - what is the purpose of this line? The page is loaded `before` the scenario as well.

File: app/jobs/project_checked_on_time_job.rb [hunk 14]
```diff
@@ -8,18 +8,23 @@ def initialize(project_check_id)
...
   def perform
-    if overdue?
-      ProjectChecks::HandleOverdue.new(project_check, days_diff).call
-    elsif notify?
-      ProjectChecks::HandleNotificationDay.new(project_check, days_diff).call
+    if handler = select_handler
```
Comment (jedi1156) [inline app/jobs/project_checked_on_time_job.rb]: Why have = in condition? It is rather hard to read. How about:
```
handler = select_handler
return if handler.nil?
handler.new(project_check, policy.elapsed_days).call
```

File: app/jobs/project_checked_on_time_job.rb (hunk 11 above)
Comment (musialik) [inline app/jobs/project_checked_on_time_job.rb]: It can probably be removed now. It was added when the method was much bigger

File: app/policies/due_date_policy.rb (hunk 2 above)
Comment (musialik) [inline app/policies/due_date_policy.rb]: If I recall correctly: `due_in` `overdue?` `remind_today?` `elapsed_days`
Are you suggesting I should make the rest private? Even though someone might use them in the future?

Comment (jedi1156) [inline app/policies/due_date_policy.rb]: Having fewer public methods make its interface more compact and thus - easier to maintain and modify. If these methods are needed, then they can be made public in the future.
An exception would be if you had an interface for policy that you have to implement. i.e. this set of methods would be needed here in this form and naming for other classes. In such case, I wouldn't argue to have them all public. An example might be a class in a gem that exposes some set of documented methods to the world.

To sum up:
- If you plan to use these methods directly for some purpose in the app or have this convention (the same set of public methods) for all (or some of) the policies, then it is ok.
- If you predict that it might be useful, but you don't really know where or what for at this stage, then I suggest making the rest of methods private for the time being.

File: spec/decorators/project_check_decorator_spec.rb [hunk 15]
```diff
@@ -129,4 +129,16 @@
       it { is_expected.to eq "disabled" }
     end
   end
+
+  describe "#days_to_deadline_as_number" do
+    it "includes disabled period in the calculations" do
+      reminder = create(:reminder, valid_for_n_days: 30)
+      check = create(:project_check,
```
Comment (jedi1156) [inline spec/decorators/project_check_decorator_spec.rb]: I sugggest moving `check`, `reminder` and `decorator` to `let`s inside `"#days_to_deadline_as_number"`. Also, how about `subject { decorator.days_to_deadline_as_number }`?
Finally:
```
describe "#days_to_deadline_as_number" do
  subject { decorator.days_to_deadline_as_number }
  let(:decorator) { described_class.new(check) }
  let!(:reminder) { create(:reminder, valid_for_n_days: 30) }
  let(:check) do
    check = create(:project_check, reminder: reminder, last_check_date: 20.days.ago, last_check_date_without_disabled_period: 10.days.ago)
  end

  it "includes disabled period in the calculations" do
    is_expected.to eq 20
  end
```

File: spec/policies/due_date_policy_spec.rb [hunk 16]
```diff
@@ -0,0 +1,217 @@
+require "rails_helper"
+
+describe DueDatePolicy do
```
Comment (jedi1156) [inline spec/policies/due_date_policy_spec.rb]: If you wrote it from scratch, I would say "have only public methods tested". But you have this spec already and you have methods that potentially can be public methods some day. In this case - leave it be. If however, some of the public methods would be changed, then I suggest aiming at testing only public methods.
